source/extensions/access_loggers/grpc/grpc_access_log_impl.h

```diff
@@ -17,24 +19,191 @@ namespace Extensions {
 namespace AccessLoggers {
 namespace GrpcCommon {
 
+static constexpr absl::string_view GRPC_LOG_STATS_PREFIX = "access_logs.grpc_access_log.";
+
+#define CRITICAL_ACCESS_LOGGER_GRPC_CLIENT_STATS(COUNTER, GAUGE)                                   \
+  COUNTER(critical_logs_succeeded)                                                                 \
+  COUNTER(ciritcal_logs_disposed)                                                                  \
+  COUNTER(pending_timeout)                                                                         \
+  GAUGE(pending_critical_logs, Accumulate)
+
+struct CriticalAccessLoggerGrpcClientStats {
+  CRITICAL_ACCESS_LOGGER_GRPC_CLIENT_STATS(GENERATE_COUNTER_STRUCT, GENERATE_GAUGE_STRUCT)
+};
+
+template <typename RequestType>
+class CriticalAccessLoggerGrpcClientImpl
+    : public Common::CriticalAccessLoggerGrpcClient<RequestType> {
+public:
+  using ResponseType = envoy::service::accesslog::v3::BufferedCriticalAccessLogsResponse;
+
+  CriticalAccessLoggerGrpcClientImpl(const Grpc::RawAsyncClientSharedPtr& client,
+                                     const Protobuf::MethodDescriptor& method,
+                                     Event::Dispatcher& dispatcher, Stats::Scope& scope,
+                                     uint64_t message_ack_timeout,
+                                     uint64_t max_critical_buffer_size_bytes)
+      : client_(client), service_method_(method), dispatcher_(dispatcher),
+        message_ack_timeout_(message_ack_timeout),
+        stats_({CRITICAL_ACCESS_LOGGER_GRPC_CLIENT_STATS(
+            POOL_COUNTER_PREFIX(scope, GRPC_LOG_STATS_PREFIX.data()),
+            POOL_GAUGE_PREFIX(scope, GRPC_LOG_STATS_PREFIX.data()))}),
+        max_critical_buffer_size_bytes_(max_critical_buffer_size_bytes) {}
+
+  struct BufferedMessage {
+    enum class State {
+      Initial,
+      Pending,
+    };
+
+    Event::TimerPtr timer_;
+    State state_{State::Initial};
+    RequestType message_;
+  };
+
+  struct ActiveStream : public Grpc::AsyncStreamCallbacks<ResponseType> {
+    ActiveStream(CriticalAccessLoggerGrpcClientImpl& parent) : parent_(parent) {}
+
+    // Grpc::AsyncStreamCallbacks
+    void onCreateInitialMetadata(Http::RequestHeaderMap&) override {}
+    void onReceiveInitialMetadata(Http::ResponseHeaderMapPtr&&) override {}
+    void onReceiveMessage(std::unique_ptr<ResponseType>&& message) override {
+      const auto& id = message->id();
+
+      if (parent_.buffered_messages_.find(id) != parent_.buffered_messages_.end()) {
+        // After response wait time exceeded, the state should be Initial.
+        if (parent_.buffered_messages_.at(id).state_ != BufferedMessage::State::Pending) {
+          return;
+        }
```

Comment:
This is not an anomaly case. Described on the response the above review thread. It will be called after `message_ack_timeout` timer callback had invoked.